source/common/tcp_proxy/tcp_proxy.h

```diff
@@ -114,17 +115,25 @@ class TunnelingConfigHelperImpl : public TunnelingConfigHelper {
 public:
   TunnelingConfigHelperImpl(
       const envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig&
-          config_message)
+          config_message,
+      Formatter::FormatterPtr hostname_fmt)
       : hostname_(config_message.hostname()), use_post_(config_message.use_post()),
-        header_parser_(Envoy::Router::HeaderParser::configure(config_message.headers_to_add())) {}
-  const std::string& hostname() const override { return hostname_; }
+        header_parser_(Envoy::Router::HeaderParser::configure(config_message.headers_to_add())),
+        hostname_fmt_(std::move(hostname_fmt)) {}
+  std::string host(const StreamInfo::StreamInfo& stream_info) const override {
+    return hostname_fmt_->format(*Http::StaticEmptyHeaders::get().request_headers,
+                                 *Http::StaticEmptyHeaders::get().response_headers,
+                                 *Http::StaticEmptyHeaders::get().response_trailers, stream_info,
+                                 absl::string_view());
+  }
   bool usePost() const override { return use_post_; }
   Envoy::Http::HeaderEvaluator& headerEvaluator() const override { return *header_parser_; }
 
 private:
   const std::string hostname_;
   const bool use_post_;
   std::unique_ptr<Envoy::Router::HeaderParser> header_parser_;
+  Formatter::FormatterPtr hostname_fmt_;
```

Comment:
Make this `const Formatter::FormatterPtr`.

Reply:
I can't, because `fromProtoConfig` does not return `const`.

Reply:
You can if you initialize this in initialization list and extract
```
envoy::config::core::v3::SubstitutionFormatString substitution_format_config;
  substitution_format_config.mutable_text_format_source()->set_inline_string(
      config_message.hostname());
  return Formatter::SubstitutionFormatStringUtils::fromProtoConfig(
      substitution_format_config, context);
```
to a helper function.